internal/backup/csi_volumesnapshot.go

```diff
@@ -0,0 +1,120 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	snapshotv1beta1api "github.com/kubernetes-csi/external-snapshotter/v2/pkg/apis/volumesnapshot/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/vmware-tanzu/velero-plugin-for-csi/internal/util"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+)
+
+// VolumeSnapshotBackupItemAction is a backup item action plugin to backup
+// CSI VolumeSnapshot objects using Velero
+type VolumeSnapshotBackupItemAction struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns information indicating that the VolumeSnapshotBackupItemAction should be invoked to backup volumesnapshots.
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+func setVolumeSnapshotAnnotationsAndLabels(vs *snapshotv1beta1api.VolumeSnapshot, vals map[string]string) {
+	if vs.Annotations == nil {
+		vs.Annotations = make(map[string]string)
+	}
+	if vs.Labels == nil {
+		vs.Labels = make(map[string]string)
+	}
+
+	for k, v := range vals {
+		vs.Annotations[k] = v
+		vs.Labels[k] = v
+	}
+}
+
+// Execute backs VolumeSnapshotContents associated with the volumesnapshot as additional resource.
+func (p *VolumeSnapshotBackupItemAction) Execute(item runtime.Unstructured, backup *velerov1api.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	p.Log.Infof("Executing VolumeSnapshotBackupItemAction")
+
+	var vs snapshotv1beta1api.VolumeSnapshot
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &vs); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	_, snapshotClient, err := util.GetClients()
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+	p.Log.Infof("Getting VolumesnapshotContent for Volumesnapshot %s/%s", vs.Namespace, vs.Name)
+	vsc, err := util.GetVolumeSnapshotContentForVolumeSnapshot(&vs, snapshotClient.SnapshotV1beta1(), p.Log)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	vals := map[string]string{
+		util.VolumeSnapshotHandleLabel: *vsc.Status.SnapshotHandle,
+		util.CSIDriverNameLabel:        vsc.Spec.Driver,
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// Capture the vsc's deletesnapshot secret annotations into the backup of the volumesnapshot
+	if util.IsVolumeSnapshotContentHasDeleteSecret(vsc) {
```

Comment:
Similar to the previous comment - Should this block of logic move into its own backup item action, that runs for VolumeSnapshotContent? That way any time we back up a VolumeSnapshotContent, we back up its related secret. Also keeps this plugin simpler.

Reply:
The current implementation doesn't explicitly add the VolumeSnapshotContent to the backup. But, now that I think about this, it might be better to have a restore item action plugin for VolumeSnapshotContents that does the conversion from dynamic snapshot to static snapshot.